sycl/include/syclcompat/math.hpp

```diff
@@ -0,0 +1,1052 @@
+  return compare(a[0], b[0], binary_op) && compare(a[1], b[1], binary_op);
+}
+
+/// Performs 2 element unordered comparison and return true if both results are
+/// true.
+/// \param [in] a The first value
+/// \param [in] b The second value
+/// \param [in] binary_op functor that implements the binary operation
+/// \returns the comparison result
+template <typename ValueT, class BinaryOperation>
+inline std::enable_if_t<ValueT::size() == 2, bool>
+unordered_compare_both(const ValueT a, const ValueT b,
+                       const BinaryOperation binary_op) {
+  return unordered_compare(a[0], b[0], binary_op) &&
+         unordered_compare(a[1], b[1], binary_op);
+}
+
+/// Performs 2 elements comparison, compare result of each element is 0 (false)
+/// or 0xffff (true), returns an unsigned int by composing compare result of two
+/// elements.
+/// \param [in] a The first value
+/// \param [in] b The second value
+/// \param [in] binary_op functor that implements the binary operation
+/// \returns the comparison result
+template <typename ValueT, class BinaryOperation>
+inline unsigned compare_mask(const sycl::vec<ValueT, 2> a,
+                             const sycl::vec<ValueT, 2> b,
+                             const BinaryOperation binary_op) {
+  return sycl::vec<short, 2>(-compare(a[0], b[0], binary_op),
+                             -compare(a[1], b[1], binary_op))
+      .as<sycl::vec<unsigned, 1>>();
+}
+
+/// Performs 2 elements unordered comparison, compare result of each element is
+/// 0 (false) or 0xffff (true), returns an unsigned int by composing compare
+/// result of two elements.
+/// \param [in] a The first value
+/// \param [in] b The second value
+/// \param [in] binary_op functor that implements the binary operation
+/// \returns the comparison result
+template <typename ValueT, class BinaryOperation>
+inline unsigned unordered_compare_mask(const sycl::vec<ValueT, 2> a,
+                                       const sycl::vec<ValueT, 2> b,
+                                       const BinaryOperation binary_op) {
+  return sycl::vec<short, 2>(-unordered_compare(a[0], b[0], binary_op),
+                             -unordered_compare(a[1], b[1], binary_op))
+      .as<sycl::vec<unsigned, 1>>();
+}
+
+/// Determine whether 2 element value is NaN.
+/// \param [in] a The input value
+/// \returns the comparison result
+template <typename ValueT>
+inline std::enable_if_t<ValueT::size() == 2, ValueT> isnan(const ValueT a) {
+  return {detail::isnan(a[0]), detail::isnan(a[1])};
+}
+
+/// cbrt function wrapper.
+template <typename ValueT> inline ValueT cbrt(ValueT val) {
+  return sycl::cbrt((ValueT)val);
+}
+
+// min function overloads.
+// For floating-point types, `float` or `double` arguments are acceptable.
+// For integer types, `std::uint32_t`, `std::int32_t`, `std::uint64_t` or
+// `std::int64_t` type arguments are acceptable.
+inline double min(const double a, const float b) {
```

Comment:
Hi @Alcpz , we recently found an issue in these min/max overloads.
In this implementation, we use std::int??_t/uint??_t, but these types are not fundamental types.
If user pass 2 longlong type variables to dpct::max, and if int64_t is typedef of long, then compiler does not know which overload should be chosen.
So I suggest to use fundamental types (int/long/longlong/...) here instead of fixed width integer types.

Reply:
Or is something like below better?
```
template <typename T1, typename T2>
std::enable_if_t<std::is_integral_v<T1> && std::is_integral_v<T2>,
                 decltype(T1(0) + T2(0))>
min(T1 a, T2 b) {
  return sycl::min(static_cast<decltype(a + b)>(a),
                   static_cast<decltype(a + b)>(b));
}
template <typename T1, typename T2>
std::enable_if_t<std::is_floating_point_v<T1> && std::is_floating_point_v<T2>,
                 decltype(T1(0) + T2(0))>
min(T1 a, T2 b) {
  return sycl::fmin(static_cast<decltype(a + b)>(a),
                    static_cast<decltype(a + b)>(b));
}

template <typename T1, typename T2>
std::enable_if_t<std::is_integral_v<T1> && std::is_integral_v<T2>,
                 decltype(T1(0) + T2(0))>
max(T1 a, T2 b) {
  return sycl::max(static_cast<decltype(a + b)>(a),
                   static_cast<decltype(a + b)>(b));
}
template <typename T1, typename T2>
std::enable_if_t<std::is_floating_point_v<T1> && std::is_floating_point_v<T2>,
                 decltype(T1(0) + T2(0))>
max(T1 a, T2 b) {
  return sycl::fmax(static_cast<decltype(a + b)>(a),
                    static_cast<decltype(a + b)>(b));
}
```

Reply:
Thanks for pointing this @zhiweij1. I will add the changes you propose once I find time to get back to these PRs.

Reply:
Since this PR is closed, I will ping you in the appropriate PR that contains these changes